### PR TITLE
perf: collapse palette taxonomy subqueries into JOIN+GROUP BY (#127)

### LINKED ISSUE #127: Collapse correlated subqueries in palette author/series/tag queries
## Context

The palette author/series/tag queries in `db::search_palette` each do **two passes over the same join chain**:

```sql
SELECT a.id, a.name,
       (SELECT COUNT(*) FROM books_authors_link bal
          JOIN books b ON b.id = bal.book
          JOIN libraries l ON l.id = b.library_id
         WHERE bal.author = a.id AND l.path = ?1) AS book_count   -- 1
FROM authors a
WHERE a.name LIKE ?2 ESCAPE '\\'
  AND EXISTS (SELECT 1 FROM books_authors_link bal
                JOIN books b ON b.id = bal.book
                JOIN libraries l ON l.id = b.library_id
               WHERE bal.author = a.id AND l.path = ?1)           -- 2
ORDER BY book_count DESC, a.name LIMIT 5
```

Pass 1 is the `book_count` correlated subquery; pass 2 is the `EXISTS` correlated subquery — both walk `books_authors_link → books → libraries` for every candidate author. On a multi-thousand-book library this is ~2× the work needed for the same predicate (`bal.author = a.id AND l.path = ?1`).

## Files

- `db/src/queries.rs` — `search_palette`, three queries (authors, series, tags)

## Desired outcome

Rewrite as a single `JOIN ... GROUP BY ... HAVING COUNT(*) > 0` per category. Sketch:

```sql
SELECT a.id, a.name, COUNT(*) AS book_count
FROM authors a
JOIN books_authors_link bal ON bal.author = a.id
JOIN books b ON b.id = bal.book
JOIN libraries l ON l.id = b.library_id
WHERE a.name LIKE ?2 ESCAPE '\\' AND l.path = ?1
GROUP BY a.id
HAVING book_count > 0
ORDER BY book_count DESC, a.name
LIMIT ?3
```

`HAVING` replaces the `EXISTS` predicate; `COUNT(*)` replaces the subquery. Use `EXPLAIN QUERY PLAN` to verify the planner picks a good index path.

## Measurement

The palette already returns `duration_ms`. Before/after numbers from a realistic library (e.g. seeded 5k-book test) should bound the win.

## Origin

PR #103 (search palette) Claude code-review — severity **Low** (\"easy to measure first\").

## Acceptance

- The three palette taxonomy queries each use a single JOIN+GROUP BY+HAVING form.
- `EXPLAIN QUERY PLAN` checked in via a comment or test for each query.
- `cargo test -p omnibus-db -- palette` still green.
- `duration_ms` benchmarked before/after (capture numbers in the PR description).

### PR DESCRIPTION
**Pull request by seamus-sloan**
## Summary

Collapses the three taxonomy queries in `db::search_palette` (authors / series / tags) from two correlated subqueries each into a single `JOIN … GROUP BY` form.

Before, each candidate row in `authors` / `series` / `tags` paid for two full traversals of `link_table → books → libraries`:
1. correlated `(SELECT COUNT(*) …) AS book_count`
2. correlated `EXISTS (SELECT 1 …)` predicate

After, both collapse into one INNER-JOIN pass: `COUNT(*)` replaces the count subquery, and the join chain implicitly enforces "at least one matching book" — groups with zero matches don't appear, so `EXISTS` is redundant. Library scoping (`l.path = ?1`) moves to the WHERE clause and is applied **before** aggregation so `book_count` stays library-correct.

## SQL shape change

```sql
-- before (illustrative, authors only):
SELECT a.id, a.name,
       (SELECT COUNT(*) FROM books_authors_link bal
          JOIN books b ON …
          JOIN libraries l ON …
         WHERE bal.author = a.id AND l.path = ?1) AS book_count
FROM authors a
WHERE a.name LIKE ?2 ESCAPE '\\'
  AND EXISTS (SELECT 1 FROM books_authors_link … WHERE bal.author = a.id AND l.path = ?1)
ORDER BY book_count DESC, a.name LIMIT ?3;

-- after:
SELECT a.id, a.name, COUNT(*) AS book_count
FROM authors a
JOIN books_authors_link bal ON bal.author = a.id
JOIN books b ON b.id = bal.book
JOIN libraries l ON l.id = b.library_id
WHERE a.name LIKE ?2 ESCAPE '\\' AND l.path = ?1
GROUP BY a.id, a.name
ORDER BY book_count DESC, a.name LIMIT ?3;
```

Series and tags follow the same shape against their respective link tables. The series `author_display` correlated subquery is intentionally retained — it needs `ORDER BY … LIMIT 1` over a different join, doesn't fold cleanly into the grouped row, and only runs for the up-to-`LIMIT` outer rows.

## EXPLAIN QUERY PLAN

Captured by a new structural test (`palette_taxonomy_query_plans_use_indexes`). Plan for all three (sqlite 3.45, fresh in-memory pool):

```
SEARCH l   USING COVERING INDEX sqlite_autoindex_libraries_1 (path=?)
SEARCH b   USING COVERING INDEX idx_books_library_id (library_id=?)
SEARCH bal USING COVERING INDEX sqlite_autoindex_books_authors_link_1 (book=?)
SEARCH a   USING INTEGER PRIMARY KEY (rowid=?)
USE TEMP B-TREE FOR GROUP BY
USE TEMP B-TREE FOR ORDER BY
```

The planner drives from the highly selective `libraries.path` unique index, then walks `books` by `library_id`, then the link table by its PK, then the taxonomy entity by rowid. No full SCANs on any link table — the test asserts that explicitly so a future schema/index regression fails loudly.

## Before/after numbers

Throwaway bench: in-memory pool seeded with 1 library, 2000 books, 200 authors, 20 series, 50 tags; LIKE pattern `%Name%`; 100 iterations per query in release mode.

| Query   | Before    | After    | Speedup |
|---------|-----------|----------|---------|
| authors | 87.15 ms  | 1.60 ms  | 54.6x   |
| series  |  9.22 ms  | 2.02 ms  |  4.6x   |
| tags    | 20.20 ms  | 1.91 ms  | 10.6x   |

Authors wins biggest because it has the most distinct entities (200) — each one paid for two full join chains under the old form.

## Files changed

- `db/src/queries.rs` — three taxonomy queries rewritten; two new tests added (`palette_taxonomy_counts_scoped_per_library`, `palette_taxonomy_query_plans_use_indexes`); inline comments document the new shape and the locked-in plan.

## Behavioural equivalence

The new queries return identical rows + counts for any input:
- `INNER JOIN` over the link table = "≥1 book" → matches old `EXISTS` semantics. Groups with zero matches were already excluded by the old form too.
- `l.path = ?1` filter applied before grouping = matches old per-library scoping (covered by both the existing `palette_scoped_to_library` test and the new `palette_taxonomy_counts_scoped_per_library` test which verifies all three taxonomies don't inflate counts across libraries).
- `ORDER BY book_count DESC, name LIMIT 5` unchanged.

No corner cases flipped between forms.

## Test plan

- [x] `cargo test -p omnibus-db -- palette` — 13/13 pass (10 existing + 2 new + 1 unchanged duration smoke)
- [x] `cargo test -p omnibus-db` — 198/198 pass
- [x] `cargo test -p omnibus` — 88/88 REST integration tests pass
- [x] `cargo clippy --all-targets -- -D warnings` — clean
- [x] `cargo fmt -- --check` — clean
- [x] EXPLAIN QUERY PLAN asserts no full SCAN on any of the three link tables

Closes #127

https://claude.ai/code/session_015K9CCjLAXmrXHvtC9XbLaN

---
_Generated by [Claude Code](https://claude.ai/code/session_015K9CCjLAXmrXHvtC9XbLaN)_

### CODE REVIEW

File: db/src/queries.rs [hunk 1]
```diff
@@ -1984,19 +1984,30 @@ pub async fn search_palette(
...
+    // clause; `l.path = ?1` is applied before grouping so authors whose only
+    // books live in other libraries don't appear and book_count stays
+    // library-correct (see `palette_scoped_to_library` test).
+    //
+    // EXPLAIN QUERY PLAN (sqlite 3.45+): the planner drives from the highly
+    // selective `libraries.path` unique index, then walks `books` via
+    // `idx_books_library_id`, then `books_authors_link` via its PK (`book`),
+    // then `authors` by rowid, with a TEMP B-TREE for GROUP BY/ORDER BY.
+    // The LIKE predicate filters during aggregation. No full scans.
+    // See `palette_taxonomy_query_plans_use_indexes` for the locked-in shape.
```
[the remark this answers is not in the document]
Comment (seamus-sloan) [reply db/src/queries.rs]: Trimmed in c052568 — kept the durable why (library scoping before aggregation) and a pointer to `palette_taxonomy_query_plans_use_indexes`; removed the diff narration and the EXPLAIN-plan walkthrough.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: db/src/queries.rs [hunk 2]
```diff
@@ -2054,19 +2067,17 @@ pub async fn search_palette(
     .collect();
 
     // D. Tags — substring match, scoped to library.
+    //
+    // Same single-pass shape; uses `idx_books_tags_tag`.
```
[the remark this answers is not in the document]
Comment (seamus-sloan) [reply db/src/queries.rs]: Fixed in c052568 — dropped the `idx_books_tags_tag` reference and updated the comment to describe the actual library → books → books_tags_link drive path, matching the authors/series comments.

---
_Generated by [Claude Code](https://claude.ai/code)_
